### PR TITLE
Added packagename normalization to some extensions @related: INFRA-1360

### PR DESCRIPTION
**Pull request by aponert**

### CODE REVIEW

File: functions/normalize_php_ext.pp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+	if ($real_custom_source) {
+		$real_package_name = "php${real_php_version}-${php_extension_name}"
+	} else {
+		if (member($plain_package_names, "php-${php_extension_name}")) {
+			$real_package_name = "php-${php_extension_name}"
+		} else {
+			$real_package_name = "php${real_php_version}-${php_extension_name}"
+		}
+	}
```
Comment (saschanowak) [inline functions/normalize_php_ext.pp]: I'm not 100% sure about the puppet syntax but I could imagine the condition like that:
```suggestion
	if (!$real_custom_source && member($plain_package_names, "php-${php_extension_name}")) {
		$real_package_name = "php-${php_extension_name}"
	} else {
		$real_package_name = "php${real_php_version}-${php_extension_name}"
	}
```

File: functions/normalize_php_ext.pp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		  ####################################
+		  defined(Apt::Source['netlogix-php']) or
+		  defined(Apt::Source['netlogix-php70']) or
+			defined(Apt::Source['php5-oldstable']) {
+		$custom_source_defined = true
+	} else {
+		$custom_source_defined = false
+	}
+
+	$real_custom_source = $custom_source ? {
```
Comment (saschanowak) [inline functions/normalize_php_ext.pp]:
```suggestion
	$has_custom_source = $custom_source ? {
```

File: manifests/extension/apcu/params.pp [hunk 3]
```diff
@@ -48,7 +48,7 @@
   ]
 
   if (versioncmp($php::params::major_version, "5.6") >= 0) {
-    $package  = "${php::params::non_versioned_extension_prefix}-apcu"
+    $package  = php::normalize_php_ext('apcu')
```
Comment (saschanowak) [inline manifests/extension/apcu/params.pp:51]: Do we really need to care about php5.6 or lower or could we cleanup the code and remove the condition?

**Review by saschanowak — CHANGES_REQUESTED** [review]

File: functions/normalize_php_ext.pp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		"php-tokenizer",
+		"php-uuid",
+		"php-validate",
+		"php-xdebug",
+		"php-yaml",
+		"php-zmq",
+	]
+
+	$real_php_version = $php_version ? {
+		undef => pick(lookup('profiles::web::php_version'), $facts['php_version'][0,3]),
```
Comment (liwo) [inline functions/normalize_php_ext.pp]: That adds a dependency on our code to this module

File: functions/normalize_php_ext.pp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		default => $php_version
+	}
+
+	# for this to work, this function MUST be placed somewhere AFTER the respective resource declaration
+	if (defined(Apt::Ppa['ppa:ondrej/php5-oldstable']) or
+		  defined(Apt::Ppa['ppa:ondrej/php5']) or
+		  defined(Apt::Ppa['ppa:ondrej/php5-5.6']) or
+	    defined(Apt::Ppa['ppa:ondrej/php'])) or
+		  ####################################
+		  defined(Apt::Source['netlogix-php']) or
```
Comment (liwo) [inline functions/normalize_php_ext.pp]: Again, this adds a direct dependency on our code

File: manifests/extension/apcu/params.pp (hunk 3 above)
Comment (liwo) [reply manifests/extension/apcu/params.pp:51]: We do

**Review by liwo — CHANGES_REQUESTED** [review]

File: functions/normalize_php_ext.pp (hunk 4 above)
Comment (aponert) [reply functions/normalize_php_ext.pp]: Which is only used by us. Do you really think we will ever push our version upstream? The other way round, putting that function to netlogix module in global did not work. the function was not found in the php module then

File: functions/normalize_php_ext.pp [hunk 6]
```diff
@@ -48,36 +48,27 @@ function php::normalize_php_ext(String $php_extension_name, Optional[String] $ph
...
+	if (!$has_custom_source and member($plain_package_names, "php-${php_extension_name}")) {
+		$real_package_name = "php-${php_extension_name}"
 	} else {
-		if (member($plain_package_names, "php-${php_extension_name}")) {
-			$real_package_name = "php-${php_extension_name}"
-		} else {
-			$real_package_name = "php${real_php_version}-${php_extension_name}"
+		$real_package_name = "php${real_php_version}-${php_extension_name}"
+		# resources work in puppet > 4 inside functions
+		package { "php-${php_extension_name}":
```
Comment (liwo) [inline functions/normalize_php_ext.pp]: This gives a function that is used for generating a package name a huge side effect. And what happens if the function is called twice for the same extension?

**Review by liwo — CHANGES_REQUESTED** [review]

**Review by liwo — APPROVED** [review]

**Review by saschanowak — APPROVED** [review]

**Pull request merged**